### ISSUE TITLE
Provide workaround for kube-apiserver HA connection loss causing nginx-ingress to not update its endpoints for 15mins

### ISSUE THREAD

**Issue by riaan53** [rancher/rancher#19558]
**What kind of request is this (question/bug/enhancement/feature request):**
bug (but not in Rancher itself)

**Steps to reproduce (least amount of steps as possible):**
1. Create a 3 node RKE cluster with all roles, including default nginx ingress
2. Add a simple http deployment workload running on all 3 nodes
3. Do some recurring calls to the http workload from outside through the ingress.
4. Figure out to what apiserver the kubernetes default service is pointing to on the node
`sudo cat /proc/net/nf_conntrack | grep 10.43.0.1 | grep "ESTABLISHED src=<src_ip>"`
5. Kill the apiserver on the other node (shutdown or pull out network cable) or send SIGSTOP signal (did not test SIGSTOP but should work is explained in tickets below) to apiserver if its on same node
6. Monitor workload endpoints, nginx-ingress endpoints, nginx-ingress logs, and workload http response.

**Result:**
Because of issue in k8s go-client and/or other components it does not detect the tcp socket is closed and thus does not reconnect. The endpoints updates are therefore not updated on the nginx-ingress that was connected to the apiserver that was killed. Thus results in possible traffic loss until a timeout kicks in (not sure what) at +/-15mins then nginx-ingress reconnects and everything is happy again.

**Other details that may be helpful:**
We tried tweaking kernel tcp timeout settings on RancherOS but did not seem to make any difference.

The k8s go-client provides pluggable transports so might be able to do something there.

Issues open:
https://github.com/kubernetes/ingress-nginx/issues/3985
https://github.com/kubernetes/client-go/issues/374
https://github.com/kubernetes/kubernetes/issues/65012

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.1
- Installation option (single install/HA): Rancher HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): imported
- Machine type (cloud/VM/metal) and specifications (CPU/memory): vmware
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T22:29:25Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:30:26Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
[rancher@rancheros-s-4vcpu-8gb-fra1-03 ~]$ docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:20:43 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:28:38 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Comment by deniseschannon** [comment]
Just to clarify, you aren't running rancher at all. It's just a RKE cluster and normal workloads on it? If so, I'd like to transfer this to rancher/rke repo which handles RKE CLI issues.

**Comment by riaan53** [comment]
Hi, the cluster is rke created (latests) and then rancher installed via helm. This issue is present in most k8s clusters and manifest in components that use the standard k8s go-client like nginx ingress controller. Feel free to move it to rke repo.

**Comment by axeal** [comment]
The issue (https://github.com/kubernetes/client-go/issues/374) was mitigated in the Kubernetes client-go package v0.20.0 (https://github.com/kubernetes/kubernetes/pull/95981), which was released alongside Kubernetes v1.20. ingress-nginx updated to v0.20.0 of the client-go package in v0.42.0 (https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v0.42.0) and Rancher ships ingress-nginx v0.43.0 for RKE CLI/Rancher provisioned Kubernetes clusters with Kubernetes v1.20.

I reproduced the issue in a lab with Kubernetes v1.18.15-rancher1-1 / ingress-nginx v0.35.0, and then tested after upgrading to Kubernetes v1.20.5 / ingress-nginx v0.43.0 and can confirm that this mitigates the issue. An ingress-nginx controller connected to a downed controplane node times out the connection within a minute, establishing a connection to another kube-apiserver and thus correctly updating the available endpoints.
